### PR TITLE
Add Bun.Glob

### PR DESCRIPTION
**Pull request by zackradisic**
### What does this PR do?

Still WIP

This adds a `Glob` class to bun which implements glob pattern matching on filesystem/strings.

Basic example:
```typescript
import { Glob } from "bun";

const glob = new Glob("**/*.ts");
const entries = await glob.match({ cwd: "src" });

console.log("Entries", entries);
```

This adds

- [x] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

Still need to do:
- [x] symlink support
- [x] more comprehensive tests (see blow)
- [x] return async iterator for `Glob.scan()`

### How did you verify your code works?

I wrote automated tests:
- [x] Basic tests for glob matching/walking
- [x] Copy tests from fast-glob
- [x] Weird edge case tests (invalid surrogate pairs, filenames with latin1 characters, etc.)
- [x] Stress tests

If Zig files changed:
- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I or my editor ran `zig fmt` on the changed files
- [x] I included a test for the new code, or an existing test covers it

If new methods, getters, or setters were added to a publicly exposed class:
- [x] I added TypeScript types for the new methods, getters, or setters

If dependencies in tests changed:
- [x] I made sure that specific versions of dependencies are used instead of ranged or tagged versions

If \*.classes.ts files were added or changed:
- [x] I ran `make codegen` to regenerate the C++ and Zig code

### CODE REVIEW

File: src/bun.js/api/glob.zig [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+is_ascii: bool,
+
+const MatchOpts = struct {
+    cwd: ?BunString,
+    dot: bool,
+    absolute: bool,
+
+    fn fromJS(globalThis: *JSGlobalObject, arguments: *ArgumentsSlice, comptime fnName: []const u8, arena: *Arena) ?MatchOpts {
+        const optsObj: JSValue = arguments.nextEat() orelse return null;
+        if (!optsObj.isObject()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig:53]: Probably should ignore if the value is `undefined`

File: src/bun.js/api/glob.zig [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            return null;
+        }
+
+        var out: MatchOpts = .{
+            .cwd = null,
+            .dot = false,
+            .absolute = false,
+        };
+
+        if (optsObj.get(globalThis, "absolute")) |absoluteVal| parse_absolute: {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig]: Same here, can use `getTruthy` which ignores null or undefined

File: src/bun.js/api/glob.zig [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    return BunString.toJSArray(globalThis, globWalk.matchedPaths.items[0..]);
+}
+
+fn makeGlobWalker(
+    this: *Glob,
+    globalThis: *JSGlobalObject,
+    arguments: *ArgumentsSlice,
+    comptime fnName: []const u8,
+    alloc: Allocator,
+    arena_: Arena,
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig]: The pointer to the arena will no longer be valid outside this scope. This is a very confusing thing about the arena allocator in Zig’s standard library.

File: src/glob.zig [hunk 4]
```diff
@@ -0,0 +1,2555 @@
...
+    }
+};
+
+pub const GlobWalker = struct {
+    pub const Result = Maybe(void);
+
+    arena: Arena = undefined,
+
+    /// not owned by this struct
+    pattern: []const u8 = undefined,
```
Comment (Jarred-Sumner) [inline src/glob.zig]: We have to be very careful whenever `undefined` is used 

Can this instead be an empty string?

File: src/glob.zig [hunk 5]
```diff
@@ -0,0 +1,2555 @@
+const std = @import("std");
+const math = std.math;
+const mem = std.mem;
+const BunString = @import("./bun.zig").String;
+const expect = std.testing.expect;
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayListUnmanaged;
+const ArrayListManaged = std.ArrayList;
+const DirIterator = @import("./bun.js/node/dir_iterator.zig");
+const Syscall = @import("./sys.zig");
```
Comment (Jarred-Sumner) [inline src/glob.zig]: This should really be encoded in a linter, but can you use namespaces instead of relative imports when possible? This makes it easier for us to move files around in the future and sometimes also with autocomplete 

`bun.sys` in this case

File: src/glob.zig [hunk 6]
```diff
@@ -0,0 +1,2555 @@
...
+        this: *GlobWalker,
+        err: Syscall.Error,
+        path_buf: [:0]const u8,
+    ) Syscall.Error {
+        @memcpy(this.pathBuf[0 .. path_buf.len + 1], @as([]const u8, @ptrCast(path_buf[0 .. path_buf.len + 1])));
+        return err.withPath(this.pathBuf[0 .. path_buf.len + 1]);
+    }
+
+    pub fn walk(this: *GlobWalker) !Maybe(void) {
+        if (this.patternComponents.items.len == 0) return .{ .result = undefined };
```
Comment (Jarred-Sumner) [inline src/glob.zig]: Why are we returning undefined? Lets use an empty string or something which indicates it is empty instead of undefined. Undefined is undefined behavior in release builds

File: src/glob.zig [hunk 7]
```diff
@@ -0,0 +1,2555 @@
...
+
+        while (this.workbuf.items.len > 0) {
+            const work_item: WorkItem = this.workbuf.pop();
+            switch (try this.handleDirEntries(&work_item, &path_buf, false)) {
+                .err => |err| return .{ .err = err },
+                else => {},
+            }
+        }
+
+        return .{ .result = undefined };
```
Comment (Jarred-Sumner) [inline src/glob.zig]: Same here

File: src/glob.zig [hunk 8]
```diff
@@ -0,0 +1,2555 @@
...
+                    this.patternComponents.items[component_idx + 1].syntax_hint == .Double) : (component_idx += 1)
+                {}
+            }
+            break :component_idx component_idx;
+        };
+
+        var pattern = this.patternComponents.items[component_idx];
+        // On windows we convert the codepoint buffer upfront so no need for
+        // these checks
+        if (comptime !isWindows) {
```
Comment (Jarred-Sumner) [inline src/glob.zig]: Can you add an assert for this on Windows?

File: src/bun.js/api/glob.zig [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+                if (ResolvePath.Platform.auto.isAbsolute(cwd_str_raw.slice())) {
+                    const cwd_str = cwd_str_raw.clone(arena.allocator()) catch {
+                        globalThis.throwOutOfMemory();
+                        return null;
+                    };
+                    break :cwd_str cwd_str.ptr[0..cwd_str.len];
+                }
+
+                var path_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+                var path_buf2: [bun.MAX_PATH_BYTES]u8 = undefined;
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig]: I suggest making this `bun.MAX_PATH_BYTES*2` just in case

File: src/bun.js/api/glob.zig [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+                const cwd = std.os.getcwd(&path_buf) catch {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig]: Use `bun.sys.getcwd` instead because `getcwd` behaves strangely on Windows iirc

File: src/bun.js/api/glob.zig [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+    }
+
+    return glob;
+}
+
+pub fn finalize(
+    this: *Glob,
+) callconv(.C) void {
+    const alloc = JSC.VirtualMachine.get().allocator;
+    alloc.free(this.pattern);
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig:386]: What keeps the `pattern` alive if the `Glob` object is freed before the `Promise` for an async task resolves successfully? You might need to use `hasPendingActivity` and make it return `true` when promises have not been fulfilled yet.

Comment (Jarred-Sumner) [reply src/bun.js/api/glob.zig:386]: Something like this should reproduce it:

```zig
import {Glob} from 'bun';
const promise = (async () => () {
  const glob = new Glob("**/*.zig")
  return glob.match("/");
})();
Bun.gc(true);
await promise;
```

File: bench/glob/match.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
+import { run, bench, group } from "mitata";
+import fg from "fast-glob";
+import { Glob, Transpiler } from "bun";
```
Comment (Jarred-Sumner) [inline bench/glob/match.js]: maybe we should have a lowercase `glob` API to avoid the constructor

```js
import {glob} from 'bun';
glob.match("pattern", "folder')
```

File: src/glob.zig [hunk 13]
```diff
@@ -0,0 +1,2555 @@
...
+                needles[i] = @splat(tok);
+            }
+            break :needles needles;
+        };
+
+        var i: usize = 0;
+        while (i + 16 <= pattern.len) : (i += 16) {
+            const haystack: @Vector(16, u8) = pattern[i..][0..16].*;
+            inline for (needles) |needle| {
+                if (std.simd.firstTrue(needle == haystack) != null) return true;
```
Comment (Jarred-Sumner) [inline src/glob.zig]: nice

File: src/glob.zig [hunk 14]
```diff
@@ -0,0 +1,2555 @@
...
+                        }
+                        continue;
+                    };
+
+                    const subdir_parts: []const []const u8 = &[_][]const u8{
+                        dir_path[0..dir_path.len],
+                        entry_name,
+                    };
+
+                    const subdir_entry_name = try std.fs.path.join(this.arena.allocator(), subdir_parts);
```
Comment (Jarred-Sumner) [inline src/glob.zig]: Can we use bun's file path join logic instead of `std.fs.path`? I don't generally trust the zig standard library for string-parsing related code in it's current level of stability. It would be unsurprising if it didn't work with Windows' UTF-16 ill-formed surrogates or otherwise handle unicode correctly.

File: src/glob.zig [hunk 15]
```diff
@@ -0,0 +1,2555 @@
...
+    }
+
+    fn matchPatternImpl(
+        this: *GlobWalker,
+        pattern_component: *Component,
+        filepath: []const u8,
+    ) bool {
+        // FIXME: handle utf-16?
+        if (!this.dot and GlobWalker.startsWithDot(filepath)) return false;
+        switch (pattern_component.syntax_hint) {
```
Comment (Jarred-Sumner) [inline src/glob.zig]: Small style thing: `switch` can be an expression

```suggestion
return switch (pattern_component.syntax_hint) {
            .Double, .Single => true,
            .WildcardFilepath => if (comptime !isWindows)
                matchWildcardFilepath(this.pattern[pattern_component.start .. pattern_component.start + pattern_component.len], filepath)
            else
                this.matchPatternSlow(pattern_component, filepath),
            .Literal => if (comptime !isWindows)
                matchWildcardLiteral(this.pattern[pattern_component.start .. pattern_component.start + pattern_component.len], filepath)
            else
                this.matchPatternSlow(pattern_component, filepath),
            else => this.matchPatternSlow(pattern_component, filepath),
};
```

File: src/glob.zig [hunk 16]
```diff
@@ -0,0 +1,2555 @@
...
+        this: *GlobWalker,
+        entry_name: []const u8,
+        component_idx: u32,
+        is_last: bool,
+        pattern: *Component,
+        next_pattern: ?*Component,
+    ) bool {
+        // Handle case a)
+        if (!is_last) return pattern.syntax_hint == .Double and
+            component_idx + 1 == this.patternComponents.items.len - 1 and
```
Comment (Jarred-Sumner) [inline src/glob.zig]:
```suggestion
            component_idx + 1 == this.patternComponents.items.len -| 1 and
```

File: src/bun.js/api/glob.zig (hunk 3 above)
Comment (zackradisic) [reply src/bun.js/api/glob.zig]: What is the best way to handle this, because the GlobWalker wants to own an `Arena`. Should I pass an `*Arena` to this function and then dereference it to give it to the `GlobWalker`?

Or should `GlobWalker` take an `*Arena`, and only when it is done allocating it can dereference it and store it?

File: src/glob.zig (hunk 6 above)
Comment (zackradisic) [reply src/glob.zig]: I changed it to `u0` is that okay?

**Comment by The-King-of-Toasters** [comment]
FWIW, I've updated https://github.com/The-King-of-Toasters/globlin with your syntax changes (which I should've done a while ago). Just some questions:

- Why strip out the micromatch test suite?
- What is the purpose of `preprocess_glob`?

Also a shoutout in the release notes would be nice ;).

**Comment by zackradisic** [comment]
> * Why strip out the micromatch test suite?
> * What is the purpose of `preprocess_glob`?



Zig tests are not working in bun right now, and I made some slight changes that caused all the tests to fail to compile, so I commented them out for now, but I'm going to port them to Typescript in globMatch.test.ts

To be honest I can't exactly remember what I had in mind for `preprocess_glob()`, I think the idea was to compile a glob pattern into tokens (e.g. `**` becomes `Globstar`) or something

> Also a shoutout in the release notes would be nice ;).

For sure!

File: test/js/bun/glob/scan.test.ts [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+  test("invalid surrogate pairs", async () => {
+    const pattern = `**/*.{md,\uD83D\uD800}`;
+    const cwd = "test/js/bun/glob";
+
+    const glob = new Glob(pattern);
+    const entries = await Array.fromAsync(glob.scan({ cwd }));
+
+    expect(entries.sort()).toEqual(
+      [
+        "fixtures/file.md",
```
Comment (Electroid) [inline test/js/bun/glob/scan.test.ts:147]: For the fixtures, instead of having them committed to source, I would recommend importing `createTmpDir` (not exact signature) from `"harness"`. This makes a bunch of temp files in a temp directory.

Comment (zackradisic) [reply test/js/bun/glob/scan.test.ts:147]: Removed the fixtures from git also the tests need to scan the `test/js/bun/glob` dir so I made a function to create a temp `fixtures` dir there

File: test/js/bun/glob/stress.test.ts [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+];
+
+test("Glob.scan stress test", async () => {
+  await Promise.all(
+    Array(1000)
+      .fill(null)
+      .map(() =>
+        Array.fromAsync(new Glob("src/**/*.zig").scan()).then(results => {
+          const set = new Set(results);
+          return set.size == paths.length && paths.every(path => set.has(path));
```
Comment (Jarred-Sumner) [inline test/js/bun/glob/stress.test.ts:31]: I am confused why it would be matching `*.zig` for a list of `.md` and `.json` files.

File: test/js/bun/glob/match.test.ts [hunk 19]
```diff
@@ -0,0 +1,1501 @@
...
+    glob = new Glob("hello/*/friends");
+    expect(glob.match("hello/\uD800\uD800/friends")).toBeTrue();
+
+    glob = new Glob("*.{js,\uD83D\u0027}");
+    expect(glob.match("runtime.node.pre.out.ts")).toBeFalse();
+    expect(glob.match("runtime.node.pre.out.js")).toBeTrue();
+  });
+
+  /**
+   * These tests are ported from micromatch, glob-match, globlin
```
Comment (Jarred-Sumner) [inline test/js/bun/glob/match.test.ts:136]: can you put their licenses at the top of this file?

File: src/bun.js/api/glob.zig [hunk 20]
```diff
@@ -0,0 +1,455 @@
...
+    defer arguments.deinit();
+
+    var arena = std.heap.ArenaAllocator.init(alloc);
+    var globWalker = this.makeGlobWalker(globalThis, &arguments, "scan", alloc, &arena) orelse {
+        arena.deinit();
+        return .undefined;
+    };
+
+    updateHasPendingActivityFlag(&this.has_pending_activity, true);
+    var task = WalkTask.create(globalThis, alloc, globWalker, &this.has_pending_activity) catch {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/glob.zig:422]: what if there are multiple simultaneous walkers? should `has_pending_activity` be a number that gets incremented/decremented?

File: src/js/builtins/Glob.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+interface Glob {
+  __scan(opts);
```
Comment (Jarred-Sumner) [inline src/js/builtins/Glob.ts]: this is not great, people might try to use it as an API

File: src/glob.zig [hunk 22]
```diff
@@ -0,0 +1,3021 @@
...
+            .cursor = CodepointIterator.Cursor{
+                .i = this.cursor.i + i,
+                .c = @truncate(nextCp),
+                .width = 1,
+            },
+        };
+    }
+};
+
+var dummyu32 = [_]u32{ 105, 102, 32, 121, 111, 117, 32, 115, 101, 101, 32, 116, 104, 105, 115, 44, 32, 122, 97, 99, 107, 32, 105, 115, 32, 97, 110, 32, 105, 100, 105, 111, 116 };
```
Comment (Jarred-Sumner) [inline src/glob.zig]: what is this? why is it a var?

File: src/glob.zig [hunk 23]
```diff
@@ -0,0 +1,3021 @@
...
+    /// If the pattern contains "./" or "../"
+    has_relative_components: bool = false,
+
+    patternComponents: ArrayList(Component) = .{},
+    matchedPaths: ArrayList(BunString) = .{},
+    i: u32 = 0,
+
+    dot: bool = false,
+    absolute: bool = false,
+    cwd: BunString = undefined,
```
Comment (Jarred-Sumner) [inline src/glob.zig]: can this be `BunString.empty` instead? `undefined` has caused strange issues many times
```suggestion
    cwd: BunString = BunString.empty,
```

File: src/glob.zig [hunk 24]
```diff
@@ -0,0 +1,3021 @@
...
+    follow_symlinks: bool = false,
+    error_on_broken_symlinks: bool = false,
+    only_files: bool = true,
+
+    root_cwd_slice: []const u8 = undefined,
```
Comment (Jarred-Sumner) [inline src/glob.zig]: can this be `""` instead? `undefined` has caused strange issues many times
```suggestion
    root_cwd_slice: []const u8 = "",
```

File: src/glob.zig [hunk 25]
```diff
@@ -0,0 +1,3021 @@
...
+            dot,
+            absolute,
+            follow_symlinks,
+            error_on_broken_symlinks,
+            only_files,
+        );
+        return .{ .result = globWalker };
+    }
+
+    pub fn convertUtf8ToCodepoints(codepoints: []u32, pattern: []const u8) void {
```
Comment (Jarred-Sumner) [inline src/glob.zig]: technically on Windows, this can be literally anything unfortunately.

File: src/glob.zig [hunk 26]
```diff
@@ -0,0 +1,3021 @@
...
+    ) Syscall.Error {
+        @memcpy(this.pathBuf[0 .. path_buf.len + 1], @as([]const u8, @ptrCast(path_buf[0 .. path_buf.len + 1])));
+        return err.withPath(this.pathBuf[0 .. path_buf.len + 1]);
+    }
+
+    pub fn walk(this: *GlobWalker) !Maybe(u0) {
+        if (this.patternComponents.items.len == 0) return .{ .result = 0 };
+        var path_buf: [bun.MAX_PATH_BYTES]u8 = std.mem.zeroes([bun.MAX_PATH_BYTES]u8);
+
+        const root_path = this.cwd.toZigString().slice();
```
Comment (Jarred-Sumner) [inline src/glob.zig]: this is going to leak memory or not be thread safe (must call `.toThreadSafe` before `BunString` can be accessed on another thread)

you should copy the directory as a string IMO

File: src/glob.zig [hunk 27]
```diff
@@ -0,0 +1,3021 @@
...
+            }
+        }
+
+        dir_path.len = len;
+
+        return component_idx;
+    }
+
+    // NOTE you must check that the pattern at `idx` has `syntax_hint == .Double` first
+    fn collapseSuccessiveDoubleWildcards(this: *GlobWalker, idx: u32) u32 {
```
Comment (Jarred-Sumner) [inline src/glob.zig]: nice

File: src/glob.zig [hunk 28]
```diff
@@ -0,0 +1,3021 @@
...
+
+        const flags = std.os.O.DIRECTORY | std.os.O.RDONLY;
+        const fd = switch (Syscall.openat(cwd_fd, dir_path, flags, 0)) {
+            .err => |err| return .{
+                .err = this.handleSysErrWithPath(err, dir_path),
+            },
+            .result => |fd_| fd_,
+        };
+        defer {
+            // _ = Syscall.close(fd);
```
Comment (Jarred-Sumner) [inline src/glob.zig]:
```suggestion

```

File: src/glob.zig [hunk 29]
```diff
@@ -0,0 +1,3021 @@
...
+        switch (try this.handleDirEntriesImpl(fd, component_idx, dir_path)) {
+            .err => |err| return .{ .err = err },
+            .result => return .{ .result = 0 },
+        }
+    }
+
+    fn handleDirEntriesImpl(this: *GlobWalker, fd: bun.FileDescriptor, component_idx: u32, dir_path: [:0]const u8) !Maybe(u0) {
+        const only_files = this.only_files;
+        const dot = this.dot;
+        _ = dot;
```
Comment (Jarred-Sumner) [inline src/glob.zig]: is this intentional?

File: src/glob.zig [hunk 30]
```diff
@@ -0,0 +1,3021 @@
...
+                return true;
+        }
+
+        return false;
+    }
+
+    /// NOTE This doesn't check that there is leading dot, use `hasLeadingDot()` to do that
+    fn removeLeadingDot(filepath: []const u8, comptime allow_non_utf8: bool) []const u8 {
+        if (comptime bun.Environment.isWindows and allow_non_utf8) {
+            return filepath[4..];
```
Comment (Jarred-Sumner) [inline src/glob.zig]: can you assert that there is a leading dot?

File: src/glob.zig [hunk 31]
```diff
@@ -0,0 +1,3021 @@
...
+        } else {
+            return filepath[2..];
```
Comment (Jarred-Sumner) [inline src/glob.zig]: can you assert that there is a leading dot?

File: src/glob.zig [hunk 32]
```diff
@@ -0,0 +1,3021 @@
...
+                switch (c) {
+                    '*', '[', '{', '?', '!' => return true,
+                    else => {},
+                }
+            }
+            return false;
+        }
+
+        const syntax_tokens = comptime [_]u8{ '*', '[', '{', '?', '!' };
+        const needles: [syntax_tokens.len]@Vector(16, u8) = needles: {
```
Comment (Jarred-Sumner) [inline src/glob.zig]: can this be `comptime`?
```suggestion
        const needles: [syntax_tokens.len]@Vector(16, u8) = comptime needles: {
```

File: src/glob.zig [hunk 33]
```diff
@@ -0,0 +1,3021 @@
...
+        const iter = CodepointIterator.init(pattern);
+        var cursor = CodepointIterator.Cursor{};
+
+        var cp_len: u32 = 0;
+        var prevIsBackslash = false;
+        while (iter.next(&cursor)) : (cp_len += 1) {
+            const c = cursor.c;
+
+            const cu8: u8 = @truncate(c);
+            _ = cu8;
```
Comment (Jarred-Sumner) [inline src/glob.zig]: did you mean to keep this?
```suggestion

```

File: src/glob.zig [hunk 34]
```diff
@@ -0,0 +1,3021 @@
...
+    var negated = false;
+    while (state.glob_index < glob.len and glob[state.glob_index] == '!') {
+        negated = !negated;
+        state.glob_index += 1;
+    }
+
+    while (state.glob_index < glob.len or state.path_index.cursor.i < path.len) {
+        if (state.glob_index < glob.len) {
+            const gu8: u8 = @truncate(glob[state.glob_index]);
+            _ = gu8;
```
Comment (Jarred-Sumner) [inline src/glob.zig]: did you mean to keep this?
```suggestion

```

File: src/glob.zig [hunk 35]
```diff
@@ -0,0 +1,3021 @@
...
+            const pu8: u8 = @truncate(state.path_index.cursor.c);
+            _ = pu8;
```
Comment (Jarred-Sumner) [inline src/glob.zig]: did you mean to keep this?
```suggestion

```

File: src/js/builtins/Glob.ts (hunk 21 above)
Comment (Electroid) [reply src/js/builtins/Glob.ts]: We should make this private somehow, people _will_ use it as an API.

Comment (zackradisic) [reply src/js/builtins/Glob.ts]: We could probably make it private with the `#` syntax so we have `#__scan` which is only accessible within the class

Comment (Jarred-Sumner) [reply src/js/builtins/Glob.ts]: We can use private symbols for this. `vm.commonIdentifiers.match` and `vm.commonIdentifiers.test` I think will work.

File: test/js/bun/glob/stress.test.ts [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+  "test/js/bun/glob/fixtures/second/nested/directory/file.md",
+  "test/js/bun/glob/fixtures/third/library/b/book.md",
+  "test/js/bun/glob/fixtures/third/library/a/book.md",
+  "test/js/bun/glob/fixtures/first/file.md",
+  "test/js/bun/glob/fixtures/first/nested/file.md",
+  "test/js/bun/glob/fixtures/first/nested/directory/file.md",
+  "test/js/bun/glob/fixtures/first/nested/directory/file.json",
+];
+
+tempFixturesDir();
```
Comment (Jarred-Sumner) [inline test/js/bun/glob/stress.test.ts]: This code should not be run in the top-level of the file incase the file is loaded but the test is skipped. Instead, it should be within a `beforeAll` callback:

```suggestion
beforeAll(() => {
  tempFixturesDir();
 });
```

File: test/js/bun/glob/stress.test.ts [hunk 37]
```diff
@@ -0,0 +1,39 @@
+import { expect, test, describe } from "bun:test";
```
Comment (Jarred-Sumner) [inline test/js/bun/glob/stress.test.ts]:
```suggestion
import { expect, test, describe, beforeAll } from "bun:test";
```

File: test/js/bun/glob/stress.test.ts [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+          const set = new Set(results);
+          return set.size == paths.length && paths.every(path => set.has(path));
+        }),
+      ),
+  );
+});
+
+test("Glob.match stress test", () => {
+  for (let i = 0; i < 10000; i++) {
+    new Glob("src/**/*.zig").match("src/cli/package_manager_command.zig");
```
Comment (Jarred-Sumner) [inline test/js/bun/glob/stress.test.ts]:
```suggestion
    if (!new Glob("src/**/*.zig").match("src/cli/package_manager_command.zig")) {
       throw new Error("test failed on run " + i);
    }
```

File: src/glob.zig [hunk 39]
```diff
@@ -0,0 +1,3032 @@
...
+        }
+
+        return .{ .result = 0 };
+    }
+
+    pub fn handleSymlink(
+        this: *GlobWalker,
+        work_item: *const WorkItem,
+        scratch_path_buf: *[bun.MAX_PATH_BYTES]u8,
+    ) !Maybe(u0) {
```
Comment (Jarred-Sumner) [inline src/glob.zig]: why is this a `Maybe(u0)` rather than a `Maybe(void)`? Usually `u0` is useful for things which we might want to think of like a number but disable it

File: src/glob.zig [hunk 40]
```diff
@@ -0,0 +1,3032 @@
...
+            if (comptime bun.Environment.isPosix) {
+                const m = stat_result.mode & std.os.S.IFMT;
+                switch (m) {
+                    std.os.S.IFDIR => break :kind .directory,
+                    std.os.S.IFLNK => break :kind .sym_link,
+                    std.os.S.IFREG => break :kind .file,
+                    else => {},
+                }
+            } else if (comptime bun.Environment.isWindows) {
+                @panic("TODO!");
```
Comment (Jarred-Sumner) [inline src/glob.zig]:
```suggestion
     return bun.todo(@src(), .{.result = {}});
```

File: src/glob.zig [hunk 41]
```diff
@@ -0,0 +1,3032 @@
...
+        if (encountered_dot_dot) {
+            const flags = std.os.O.DIRECTORY | std.os.O.RDONLY;
+            const fd = switch (Syscall.openat(cwd_fd, dir_path, flags, 0)) {
+                .err => |err| return .{
+                    .err = this.handleSysErrWithPath(err, dir_path),
+                },
+                .result => |fd_| fd_,
+            };
+            defer {
+                _ = Syscall.closeAllowingStdoutAndStderr(fd);
```
Comment (Jarred-Sumner) [inline src/glob.zig]: This can just be `close`, it should never be stdout or stderr. If it is, that is a bug.
```suggestion
                _ = Syscall.close(fd);
```

File: src/glob.zig [hunk 42]
```diff
@@ -0,0 +1,3032 @@
...
+
+        const flags = std.os.O.DIRECTORY | std.os.O.RDONLY;
+        const fd = switch (Syscall.openat(cwd_fd, dir_path, flags, 0)) {
+            .err => |err| return .{
+                .err = this.handleSysErrWithPath(err, dir_path),
+            },
+            .result => |fd_| fd_,
+        };
+        defer {
+            _ = Syscall.closeAllowingStdoutAndStderr(fd);
```
Comment (Jarred-Sumner) [inline src/glob.zig]:
```suggestion
            _ = Syscall.close(fd);
```

File: test/js/bun/glob/stress.test.ts [hunk 43]
```diff
@@ -1,26 +1,26 @@
...
+  path.join(import.meta.dir, "fixtures/first/nested/directory/file.json"),
 ];
 
 beforeAll(() => {
   tempFixturesDir();
 });
 
 test("Glob.scan stress test", async () => {
-  const cwd = "test/js/bun/glob";
+  const cwd = import.meta.dir;
```
Comment (Jarred-Sumner) [inline test/js/bun/glob/stress.test.ts:23]: @zackradisic the cwd is the `test` folder in the process, not the `bun` directory. Did you manually edit bunfig.toml to make it not the `test` directory?

**Comment by Jarred-Sumner** [comment]
@zackradisic the next steps here to merge this are:

1) Add a memory leak test for sync & async which runs in a separate process, runs several scans of probably the same directory and waits for each one to complete one at a time. It should run `Bun.gc(true)` before and after, and compare `process.memoryUsage.rss()`. It should be executed with `--smol`. The memory usage gap should not be more than around 100 MB or so.
2) Fix the Arena allocator issue 
3) Fix the Windows build

I did some cleanup work to make the tests pass, fix a memory leak, do the zig upgrade

File: src/install/lockfile.zig [hunk 44]
```diff
@@ -3448,23 +3532,23 @@ pub const Package = extern struct {
...
+                // if (!strings.endsWithComptime(without_trailing_slash, "/*") and !strings.eqlComptime(without_trailing_slash, "*")) {
+                //     log.addError(source, item.loc,
+                //         \\TODO glob star * in the middle of a path. For now, try something like "packages/*", at the end of the path.
+                //     ) catch {};
+                //     continue;
+                // }
+
+                glob_workspace_paths.append(input_path) catch unreachable;
                 continue;
             } else if (strings.containsAny(input_path, "!{}[]")) {
```
Comment (dylan-conway) [inline src/install/lockfile.zig]: can we append to `glob_workspace_paths` for this case too?

Comment (dylan-conway) [reply src/install/lockfile.zig]: I think each `input_path` should be appended, because each should be a valid glob pattern

File: src/install/lockfile.zig [hunk 45]
```diff
@@ -3533,93 +3617,88 @@ pub const Package = extern struct {
             });
         }
 
-        if (asterisked_workspace_paths.items.len > 0) {
+        if (glob_workspace_paths.items.len > 0) {
```
Comment (dylan-conway) [inline src/install/lockfile.zig]: if each is appended, this could go inside the `arr.slice()` loop

File: src/install/lockfile.zig [hunk 46]
```diff
@@ -3368,6 +3376,70 @@ pub const Package = extern struct {
         version: ?[]const u8 = null,
     };
 
+    fn processWorkspaceNameGlob(
```
Comment (dylan-conway) [inline src/install/lockfile.zig]: unused?

File: src/install/lockfile.zig [hunk 47]
```diff
@@ -3368,6 +3376,70 @@ pub const Package = extern struct {
...
+        defer walker.deinit();
+
+        var iterator: GlobWalker.Iterator = .{
+            .walker = &walker,
+        };
+        switch (try iterator.init()) {
+            .err => |err| {
+                _ = err;
+                // FIXME handle errors gracefully
+                @panic("Ooops");
```
Comment (Jarred-Sumner) [inline src/install/lockfile.zig]: This does need to be done before we can ship.

File: src/install/lockfile.zig [hunk 48]
```diff
@@ -3368,6 +3376,70 @@ pub const Package = extern struct {
...
+            },
+            else => {},
+        }
+        defer iterator.deinit();
+
+        while (switch (try iterator.next()) {
+            .err => |err| {
+                _ = err;
+                // FIXME handle errors gracefully
+                @panic("Ooops");
```
Comment (Jarred-Sumner) [inline src/install/lockfile.zig]: This does need to be done before we can ship.

File: src/install/lockfile.zig [hunk 49]
```diff
@@ -3368,6 +3376,70 @@ pub const Package = extern struct {
...
+            },
+            .result => |r| r,
+        }) |matched_path| {
+            _ = matched_path;
```
Comment (Jarred-Sumner) [inline src/install/lockfile.zig]: Why isn't this returned?

File: src/install/lockfile.zig [hunk 50]
```diff
@@ -3533,93 +3617,88 @@ pub const Package = extern struct {
...
+                    false,
+                    false,
+                    false,
+                    false,
+                    false,
+                )) {
+                    .err => |err| {
+                        _ = err;
+                        // FIXME handle errors gracefully
+                        @panic("Ooops");
```
Comment (Jarred-Sumner) [inline src/install/lockfile.zig]: This does need to be handled

File: src/resolver/resolve_path.zig [hunk 51]
```diff
@@ -436,16 +436,16 @@ pub fn relativeNormalized(from: []const u8, to: []const u8, comptime platform: P
 pub fn dirname(str: []const u8, comptime platform: Platform) []const u8 {
     switch (comptime platform.resolve()) {
         .loose => {
-            const separator = lastIndexOfSeparatorLoose(str);
-            return str[0 .. separator + 1];
+            const separator = lastIndexOfSeparatorLoose(str) orelse return "";
+            return str[0..separator];
```
Comment (Jarred-Sumner) [inline src/resolver/resolve_path.zig:440]: Did you check this doesn't break anything?

**Comment by Jarred-Sumner** [comment]
🎉
